vllm/model_executor/models/molmo.py

```diff
@@ -0,0 +1,1130 @@
+):
+    processor = cached_get_processor(ctx.model_config.model, trust_remote_code=True)
+    image_processor = processor.image_processor
+
+    base_image_input_d = image_processor.image_patch_size
+    left_margin, right_margin = image_processor.overlap_margins
+    max_crops = image_processor.max_crops
+
+    # Assume: prompt_token_ids always starts with bos_token_id followed image tokens
+    max_llm_image_tokens = get_max_molmo_image_tokens(ctx)
+    if seq_len - max_llm_image_tokens - 1 < 0:
+        raise RuntimeError(
+            f"Molmo cannot process {max_crops} crops in a prompt, "
+            "please increase max_model_len or reduce number of crops"
+        )
+    
+    # The vertical image has the maximum number of image tokens due to column tokens.
+    tiling = (max_crops, 1)
+    total_margin_pixels = base_image_input_d * (right_margin + left_margin)
+    crop_patches = image_processor.base_image_input_size[0] // base_image_input_d
+    crop_window_patches = crop_patches - (right_margin + left_margin)
+    crop_window_size = crop_window_patches * base_image_input_d
+
+    h = crop_window_size * tiling[0] + total_margin_pixels
+    w = crop_window_size * tiling[1] + total_margin_pixels
+
+    dummy_image = Image.new(
+        "RGB", (w, h), color="red"
+    )
+
+    out = processor.process("dummy prompt", dummy_image)
+
+    token_ids = array(VLLM_TOKEN_ID_ARRAY_TYPE,
+        out["input_ids"][:1 + max_llm_image_tokens])
+    token_ids += array(VLLM_TOKEN_ID_ARRAY_TYPE,
+        [0]) * (seq_len - max_llm_image_tokens - 1)
+    dummy_seqdata = SequenceData(token_ids)
+    dummy_imgdata = {
+        "images": out["images"],
+        "image_input_idx": out["image_input_idx"],
+    }
+    if "image_masks" in out:
+        dummy_imgdata["image_masks"] = out["image_masks"]
+    dummy_imgdata["seq_len"] = torch.tensor(seq_len, dtype=torch.long)
+    return dummy_seqdata, {"image": dummy_imgdata}
+
+
+def pad_images(
+    max_total_crops: int,
+    images: torch.Tensor,
+    image_input_idx: torch.Tensor,
+    image_masks: Optional[torch.Tensor] = None,
+):
+    n = max_total_crops - images.shape[0]
+    images = F.pad(
+        images, (0, 0, 0, 0, 0, n), value=-1
+    )
+    image_input_idx = F.pad(
+        image_input_idx, (0, 0, 0, n), value=-1
+    )
+    if image_masks is not None:
+        image_masks = F.pad(
+            image_masks, (0, 0, 0, n), value=-1
+        )
+    return images, image_input_idx, image_masks
+
+
+def input_processor_for_molmo(ctx: InputContext, llm_inputs: LLMInputs):
+    prompt = llm_inputs["prompt"]
+    multi_modal_data = llm_inputs.get("multi_modal_data")
+    image = multi_modal_data.get("image")
+    processor = cached_get_processor(ctx.model_config.model, trust_remote_code=True)
+
+    if re.match(r"^User:[\s\S]*?(Assistant:)*$", prompt):
+        out = processor.process(prompt, image, message_format="none")
+    else:
+        out = processor.process(prompt, image)
```

Comment:
I found that this processor did not work when using the model with the OpenAI server entry point. The `llm_inputs` object receives a tokenized text prompt and the `prompt` field is `None`.

I've been tinkering and something like the code below worked:

```python
class MolmoProcessor(ProcessorMixin):
    ...
    prompt = llm_inputs["prompt"]
    multi_modal_data = llm_inputs.get("multi_modal_data")
    if multi_modal_data is None:
        image = None
    else:
        image = multi_modal_data.get("image")
    processor = cached_get_processor(ctx.model_config.model, trust_remote_code=True)

    if prompt is not None and re.match(r"^User:[\s\S]*?(Assistant:)*$", prompt):
        out = processor.process(prompt, image, message_format="none")
    elif prompt is not None:
        out = processor.process(prompt, image)
    else:
        out = processor.process(None, image, tokens=llm_inputs["prompt_token_ids"])
```

But I also needed to modify the `preprocessing_molmo.py` [file inside the huggingface checkpoint](https://huggingface.co/allenai/Molmo-72B-0924/blob/main/preprocessing_molmo.py#L127) to accept a tokenized text prompt:

```python
    def process(
        self,
        text: TextInput = None,
        images: ImageInput = None,
        *,
        tokens= None,  # <- new arg
        **kwargs: Unpack[MolmoProcessorKwargs],
    ):
        ... code before tokens definition is unchanged

        if tokens is None:
            tokens = self.get_tokens_input(
                text,
                output_kwargs["text_kwargs"]["message_format"],
                output_kwargs["text_kwargs"]["always_start_with_space"],
            )

```

Once I put these two changes together it answers a question about the image correctly.

Reply:
We should have `prompt` here for online inference - let me take a look at this issue

Reply:
On a side note, it seems not trivial to separate the logics for processing the `PIL.Image.Image` and for processing input token ids for this model, so I guess it's ok that now you define a no-op `image_input_mapper_for_molmo` and have all that logics inside `input_processor_for_molmo`, but we just need to change `preprocessing_molmo.py` for it to be able to take input token ids directly.

Reply:
Thanks for the explanation. We will update `preprocessing_molmo.py` so as to directly pass tokenized inputs.

Reply:
Thanks! I added type hints in your code and pushed it to HF.